source/fuzz/transformation_wrap_region_in_selection.cpp

```diff
@@ -0,0 +1,196 @@
+    : message_(message) {}
+
+TransformationWrapRegionInSelection::TransformationWrapRegionInSelection(
+    uint32_t region_entry_block_id, uint32_t region_exit_block_id,
+    bool branch_condition) {
+  message_.set_region_entry_block_id(region_entry_block_id);
+  message_.set_region_exit_block_id(region_exit_block_id);
+  message_.set_branch_condition(branch_condition);
+}
+
+bool TransformationWrapRegionInSelection::IsApplicable(
+    opt::IRContext* ir_context,
+    const TransformationContext& transformation_context) const {
+  // Check that it is possible to outline a region of blocks without breaking
+  // domination and structured control flow rules.
+  if (!IsApplicableToBlockRange(ir_context, message_.region_entry_block_id(),
+                                message_.region_exit_block_id())) {
+    return false;
+  }
+
+  // There must exist an irrelevant boolean constant to be used as a condition
+  // in the OpBranchConditional instruction.
+  return fuzzerutil::MaybeGetBoolConstant(ir_context, transformation_context,
+                                          message_.branch_condition(),
+                                          true) != 0;
+}
+
+void TransformationWrapRegionInSelection::Apply(
+    opt::IRContext* ir_context,
+    TransformationContext* transformation_context) const {
+  auto* new_header_block =
+      ir_context->cfg()->block(message_.region_entry_block_id());
+  assert(new_header_block->terminator()->opcode() == SpvOpBranch &&
+         "This condition should have been checked in the IsApplicable");
+
+  const auto successor_id =
+      new_header_block->terminator()->GetSingleWordInOperand(0);
+
+  // Change |entry_block|'s terminator to |OpBranchConditional|.
+  new_header_block->terminator()->SetOpcode(SpvOpBranchConditional);
+  new_header_block->terminator()->SetInOperands(
+      {{SPV_OPERAND_TYPE_ID,
+        {fuzzerutil::MaybeGetBoolConstant(ir_context, *transformation_context,
+                                          message_.branch_condition(), true)}},
+       {SPV_OPERAND_TYPE_ID, {successor_id}},
+       {SPV_OPERAND_TYPE_ID, {successor_id}}});
+
+  // Insert OpSelectionMerge before the terminator.
+  new_header_block->terminator()->InsertBefore(MakeUnique<opt::Instruction>(
+      ir_context, SpvOpSelectionMerge, 0, 0,
+      opt::Instruction::OperandList{
+          {SPV_OPERAND_TYPE_ID, {message_.region_exit_block_id()}},
+          {SPV_OPERAND_TYPE_SELECTION_CONTROL,
+           {SpvSelectionControlMaskNone}}}));
+
+  // We've change the module so we must invalidate analyses.
+  ir_context->InvalidateAnalysesExceptFor(opt::IRContext::kAnalysisNone);
+}
+
+protobufs::Transformation TransformationWrapRegionInSelection::ToMessage()
+    const {
+  protobufs::Transformation result;
+  *result.mutable_wrap_region_in_selection() = message_;
+  return result;
+}
+
+bool TransformationWrapRegionInSelection::IsApplicableToBlockRange(
```

Comment:
I wonder how much this has in common with some of the code in TransformationOutlineFunction, which is already quite similar to code in TransformationDuplicateRegionWithSelection, and for which we have a refactoring issue (#3785).

I'm fine with *not* factoring out common code here if we believe the transformations have somewhat different requirements, but it's worth considering.  Perhaps you could look at the refactoring I do when I get round to addressing #3785 and then see if you think this could make use of the functionality that I extract.  (But no need to delay this PR for that.)

Reply:
I agree that we need to refactor some common code out of those three transformations (TransformationOutlineFunction, TransformationDuplicateRegionWithSelection and this one). However, I'd prefer doing it in a separate PR (maybe as a part of #3785).

Reply:
BTW, I think some refactoring needs to be done with fuzzer passes as well.